Authentication

auloin opened the pull request:
Quite a lot already.
Need a critical look at the folder structures.
Where should I put Business logic (like email validity check)? In the model? A custom middleware in between graphql & mongoose?

Closes #2 
Closes #1

Commit:
Add .env to be ignored

Commit:
Add dotenv dependency

Commit:
Require & configure dotenv

Commit:
Configure mongoose connection

Commit:
Add users model

Commit:
Create user schema

Commit:
Add mongoose BankDetails schema

Commit:
Create mongoose Address schema

Commit:
Import Address & BankDetails schemas

Commit:
Create Address type

Commit:
Create BankDetails type

Commit:
camelCase bankDetails field

Commit:
Create User type

Commit:
Delete schema.graphql

Commit:
Add types.js as replacement for schema.graphql

Commit:
Import graphql types & remove initial typeDefs

Commit:
Add Query type with getUsers query

Commit:
Create queries resolvers file & add default getUsers resolver

Commit:
Remove set up examples & import Query resolver

Commit:
Disable mocks

Commit:
Create mutation resolvers file & add register resolver

Commit:
Add mutation type with register mutation

Commit:
Import Mutation resolver

Commit:
Add comments to delimit types & inputs

Commit:
Add AddressInput, BankDetailsInput & UserInput

Commit:
Change register parameters to UserInput

Commit:
Remove import gql from apollo-server

Commit:
Replace unique by sparse on iban field to enable nullability of key

Commit:
Destructure args to find user object

Commit:
Make return the registered user required & return it

Commit:
Add bcrypt dependencies

Commit:
Further implement register mutation

Commit:
Implement getUsers

Commit:
Remove password from user type

Commit:
Rename getUsers -> users

Commit:
Create login query

Commit:
Add jsonwebtoken dependencies & cookie-parser dependencies

Commit:
Implement login resolver

Commit:
Uninstall apollo-server dependencies

Commit:
Add apollo-server-express dependencies

Commit:
Configure express server

Commit:
Update package import

Commit:
Forward express req, res via context

Commit:
Attach the cookie to the res

Commit:
Add comment describing maxAge

Commit:
Handle the cookie to retrieve the logged user

Commit:
Add query me to retrieve the logged user

Commit:
Add logout query

Commit:
Create Reponse type & change logout return type to Response

Commit:
Remove unique/sparse properties from the models

Commit:
Add blueprintjs dependencies

Commit:
Add nextjs scripts

Commit:
Add apollo client dependencies

Commit:
Add default home page

Commit:
Prettify index.jsx

Commit:
Add with-next-apollo dependencies

Commit:
Create config file

Commit:
Rename config -> configs

Commit:
Create withApollo HOC

Commit:
Add custom <App>

Commit:
Configure App to provide Apollo to every page

Commit:
Add prettier dependencies

Commit:
Add .prettierrc

Commit:
Configure eslint

Commit:
Add & configure husky

Commit:
Modify lint-staged configuration

archived-m committed:
Add eslint dependencies

archived-m committed:
configure eslint

archived-m committed:
Run eslint --fix

archived-m committed:
Use Schema consistently

MichielLeyman commented on backend/src/graphql/resolvers/queries.js:
```diff
@@ -0,0 +1,33 @@
+const User = require('../../models/user');
+
+module.exports = {
+  users: async () => (users = await User.find()),
```
`users` is undefined. Did you want to return `User.find()`?

archived-m replied:
`users` is undefined. Did you want to return `User.find()`?

MichielLeyman commented on backend/src/graphql/resolvers/mutations.js:
```diff
@@ -0,0 +1,24 @@
+
+module.exports = {
+  register: async (_, { user }) => {
+    // TODO Valid email format
```
Are you doing validation all at once in a separate PR?
Might be worth doing straight away

archived-m replied:
Are you doing validation all at once in a separate PR?
Might be worth doing straight away

MichielLeyman commented on backend/src/index.js:
```diff
@@ -1,30 +1,53 @@
-const { ApolloServer, gql } = require('apollo-server');
+require('dotenv').config();
+const mongoose = require('mongoose');
+const express = require('express');
```
Not listed in dependencies in `package.json`

archived-m commented on backend/src/index.js:
```diff
@@ -1,30 +1,53 @@
-const { ApolloServer, gql } = require('apollo-server');
+require('dotenv').config();
+const mongoose = require('mongoose');
+const express = require('express');
```
Not listed in dependencies in `package.json`

auloin replied:
Yes there were a need to switch to apollo-server-express

archived-m commented on backend/src/models/bankDetails.js:
```diff
@@ -0,0 +1,13 @@
+
+const { Schema } = mongoose;
+
+const bankDetailsSchema = new Schema(
```
Is a constructor, bot the variable and filename should be capitalized

archived-m replied:
So right now there's one payment method, being a bank account?

archived-m commented on backend/src/models/user.js:
```diff
@@ -0,0 +1,16 @@
+
+const { Schema } = mongoose;
+
+const userSchema = new Schema({
```
Is a constructor, both the variable and filename should be capitalized

archived-m commented on backend/src/models/address.js:
```diff
@@ -0,0 +1,15 @@
+
+const { Schema } = mongoose;
+
+const addressSchema = new Schema(
```
Is a constructor, bot the variable and filename should be capitalized

archived-m commented on backend/src/graphql/resolvers/queries.js:
```diff
@@ -0,0 +1,33 @@
+    if (user) {
+      rightPassword = await bcrypt.compare(password, user.password);
+    }
+    if (!user || !rightPassword) {
```
Maybe `try` `catch` here, if it errors at any point, the email or password is incorrect.

auloin replied:
There is no error handling yet, I'll add that later.

archived-m commented on backend/src/index.js:
```diff
@@ -1,30 +1,53 @@
+    // Get the token
+    const { token } = req.cookies;
+    let user = null;
+    if (token) {
```
Can nest fewer ifs if we check for `!token` first and return a 401

auloin replied:
This mean anybody can register or login. I'm thinking of extract this logic in a separate function.

archived-m commented on backend/src/graphql/resolvers/queries.js:
```diff
@@ -0,0 +1,33 @@
+const User = require('../../models/user');
+
+module.exports = {
+  users: async () => (users = await User.find()),
```
Is this query for admins to get a listing of all users?

auloin replied:
I'll add authorization later as well as roles

archived-m commented on backend/src/graphql/resolvers/mutations.js:
```diff
@@ -0,0 +1,24 @@
+const User = require('../../models/user');
+
+module.exports = {
+  register: async (_, { user }) => {
```
Someone who freshly registers, can only create expenses right? If not, there needs to be an approve/reject mechanism

archived-m commented on backend/src/graphql/resolvers/mutations.js:
```diff
@@ -0,0 +1,24 @@
+    }
+    // Hash Password
+    const password = await bcrypt.hash(user.password, 10);
+    // Register && Return the user
```
TODO: send account confirmation email

archived-m reviewed (commented):
Great work.
Maybe it would be good to think about mailing in here already - and set up a password reset flow while we're on the topic of authentication.

If admins already have a way to get all users, why is there no way to update a user (set their role or `active` state) or delete them altogether?

archived-m reviewed (changes requested):
Also, please add a basic README on which server environment variables are required for the app to start.

archived-m committed:
Update package-lock.json

auloin commented on backend/package.json:
```diff
@@ -11,11 +11,20 @@
+    "eslint": "^5.3.0",
+    "eslint-config-airbnb": "^17.1.0",
+    "eslint-plugin-import": "^2.16.0",
+    "eslint-plugin-jsx-a11y": "^6.2.1",
```
Are react related plugins useful in the backend?

archived-m replied:
I think they're required for the airbnb style guide

Commit:
Add prettier

Commit:
Add husky & lint-staged

Commit:
Indent .eslintrc

Commit:
Fix users undefined

Commit:
Change user schemas to only require a Name instead of (First/Last)Name pair

Commit:
Update Mongoose schemas
Add trim option to every String field
Add lowercase option to email
Check email existance before saving the user

Commit:
Hash & compare passwords in the user schema

Commit:
Override eslint formatting rules by prettier

Commit:
Add prettier plugin to eslintrc

Commit:
Auto indent index.js & add express dependency

Commit:
Extract retrieval of loggedUser in a stand alone file

Commit:
Create an attemptLogin in auth.js to handle any login request

Commit:
Change Response type to Success

Commit:
Correct typo

Commit:
Move logout logic to auth.js & change logout return to Boolean

Commit:
Add JSDoc for attemptLogin & attemptLogout functions

Commit:
Add loggedIn, ensureLoggedIn & ensureLoggedOut functions to auth.js

Commit:
Create & use GuestDirective

Commit:
Create & use AuthDirective

auloin committed:
Merge pull request #6 from openknowledgebe/frontend/setup

Setup of the frontend

auloin reviewed (commented):
> Also, please add a basic README on which server environment variables are required for the app to start.

I didn't see this one. I will make sure to add a .env.example file and explain all variables in the readme.

auloin commented:
What do you think of using directives for validation & authorization? A working example of directive here 
9d0dcaa.
